libr/arch/p/arch_bf.c

```diff
@@ -0,0 +1,338 @@
+		r_strbuf_setf (&ins->esil, "%d,ptr,+=", ins->size);
+		break;
+	case '<':
+		ins->type = R_ANAL_OP_TYPE_SUB;
+		ins->size = countChar (buf, len, '<');
+		r_strbuf_setf (&ins->esil, "%d,ptr,-=", ins->size);
+		break;
+	case '+':
+		ins->size = countChar (buf, len, '+');
+		ins->type = R_ANAL_OP_TYPE_ADD;
+		r_strbuf_setf (&ins->esil, "%d,ptr,+=[1]", ins->size);
+		break;
+	case '-':
+		ins->type = R_ANAL_OP_TYPE_SUB;
+		ins->size = countChar (buf, len, '-');
+		r_strbuf_setf (&ins->esil, "%d,ptr,-=[1]", ins->size);
+		break;
+	case '.':
+		// print element in stack to screen
+		ins->type = R_ANAL_OP_TYPE_STORE;
+		r_strbuf_set (&ins->esil, "ptr,[1],scr,=[1],scr,++=");
+		break;
+	case ',':
+		ins->type = R_ANAL_OP_TYPE_LOAD;
+		r_strbuf_set (&ins->esil, "kbd,[1],ptr,=[1],kbd,++=");
+		break;
+	case 0:
+	case -1:
+		ins->type = R_ANAL_OP_TYPE_TRAP;
+		break;
+	default:
+		ins->type = R_ANAL_OP_TYPE_NOP;
+		r_strbuf_set (&ins->esil, ",");
+		break;
+	}
+	return true;
+}
+
+static bool assemble(RArch *a, RArchInstruction *ins) {
+	const char *asmstr = r_strbuf_get (&ins->code);
+	int n = 0;
+	if (asmstr[0] && asmstr[1] == ' ') {
+		asmstr += 2;
+	}
+	const char *arg = strchr (asmstr, ',');
+	const char *ref = strchr (asmstr, '[');
+	ut8 opbuf[32];
+	if (!strncmp (asmstr, "trap", 4)) {
+		if (arg) {
+			n = atoi (arg + 1);
+			memset (opbuf, 0xcc, n);
+		} else {
+			opbuf[0] = 0x90;
+			n = 1;
+		}
+	} else if (!strncmp (asmstr, "nop", 3)) {
+		if (arg) {
+			n = atoi (arg + 1);
+			memset (opbuf, 0x90, n);
+		} else {
+			opbuf[0] = 0x90;
+			n = 1;
+		}
+	} else if (!strncmp (asmstr, "inc", 3)) {
+		char ch = ref? '+': '>';
+		opbuf[0] = ch;
+		n = 1;
+	} else if (!strncmp (asmstr, "dec", 3)) {
+		char ch = ref? '-': '<';
+		opbuf[0] = ch;
+		n = 1;
+	} else if (!strncmp (asmstr, "sub", 3)) {
+		char ch = ref? '-': '<';
+		if (arg) {
+			n = atoi (arg + 1);
+			memset (opbuf, ch, n);
+		} else {
+			opbuf[0] = '<';
+			n = 1;
+		}
+	} else if (!strncmp (asmstr, "add", 3)) {
+		char ch = ref? '+': '>';
+		if (arg) {
+			n = atoi (arg + 1);
+			memset (opbuf, ch, n);
+		} else {
+			opbuf[0] = '<';
+			n = 1;
+		}
+	} else if (!strncmp (asmstr, "while", 5)) {
+		opbuf[0] = '[';
+		n = 1;
+	} else if (!strncmp (asmstr, "loop", 4)) {
+		opbuf[0] = ']';
+		n = 1;
+	} else if (!strncmp (asmstr, "in", 2)) {
+		if (arg) {
+			n = atoi (arg + 1);
+			memset (opbuf, ',', n);
+		} else {
+			opbuf[0] = ',';
+			n = 1;
+		}
+	} else if (!strncmp (asmstr, "out", 3)) {
+		if (arg) {
+			n = atoi (arg + 1);
+			memset (opbuf, '.', n);
+		} else {
+			opbuf[0] = '.';
+			n = 1;
+		}
+	}
+	r_strbuf_setbin (&ins->data, opbuf, n);
+	ins->size = n;
+	return true;
+}
+
+// --------------------
+
+static bool decode(RArch *a, RArchInstruction *ins, RArchOptions opt) {
+	bool ret = false;
+	// TODO: merge disasm+analyze into 1 function to avoid duplicating logic code. just use the opt
+	if (opt & R_ARCH_OPTION_CODE || opt & R_ARCH_OPTION_SIZE) {
```

Comment:
i'm not a huge fan of this `opt` approach.
I think this would be cleaner if each RArchPlugin would have:
```C
.size
.esil
.anal
.get_description
.opex
```

Under the hood of course those can share some code in each plugin, but from an API point of view I think it's cleaner.

Reply:
if you do this you will end up analyzing the same bytes 5 times. the reason for having a bitfield option argument is to decode the instruction ONCE and fill the structure fields that are requested. lke RAnal.op does. which resulted to make analysis 2x faster by just picking the information we need instead of everything all the time

Reply:
good point, I will think about this!